include/camp/array.hpp

```diff
@@ -0,0 +1,287 @@
+#ifndef camp_array_HPP__
+#define camp_array_HPP__
+
+#include "camp/defines.hpp"
+#include "camp/helpers.hpp"
+#include "camp/number.hpp"
+
+#include <cstddef>
+#include <stdexcept>
+#include <type_traits>
+
+namespace camp {
+   ///
+   /// Provides a portable std::array-like class.
+   /// 
+   /// Differences from std::array are listed below:
+   /// - No __device__ "at" method (exceptions are not GPU friendly)
+   /// - Calling front or back is a compile error when size() == 0
+   ///   (instead of undefined behavior at run time)
+   /// - No reverse iterators have been implemented yet
+   /// - Swap is yet to be implemented
+   ///
+   template <class T, std::size_t N>
+   struct array {
+      using value_type = T;
+      using size_type = std::size_t;
+      using difference_type = std::ptrdiff_t;
+      using reference = value_type&;
+      using const_reference = const value_type&;
+      using pointer = value_type*;
+      using const_pointer = const value_type*;
+      using iterator = pointer;
+      using const_iterator = const_pointer;
+
+      // TODO: Investigate device trap
```

Comment:
@adayton1 FWIW, we use this macro in RAJA to handle the various back-end cases https://github.com/LLNL/RAJA/blob/develop/include/RAJA/util/macros.hpp#L143

Reply:
I will look into that, but I don't want that to hold up this pull request.